### PR TITLE
Add `$type` operator

### LINKED ISSUE #355: Support `$type` element query operator
Support `$type` element query operator for all data types.

### PR DESCRIPTION
**Pull request by w84thesun**
Closes #355.

### CODE REVIEW

File: integration/query_element_test.go [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+		name, tc := name, tc
+		t.Run(name, func(t *testing.T) {
+			t.Parallel()
+
+			var actual []bson.D
+			q := bson.D{{"value", bson.D{{"$type", tc.v}}}}
+			cursor, err := collection.Find(ctx, q, options.Find().SetSort(bson.D{{"_id", 1}}))
+			if tc.err != nil {
+				require.Nil(t, tc.expectedIDs)
+				require.Equal(t, tc.err, err)
```
Comment (AlekSi) [inline integration/query_element_test.go:192]: Use `assertEqualError` helper: https://github.com/FerretDB/FerretDB/blob/252f8f6832599fe7123a3ae50c121c3bb2b30435/integration/helpers.go#L109-L122

File: internal/handlers/common/filter.go [hunk 2]
```diff
@@ -511,3 +520,186 @@ func filterFieldExprExists(fieldExist bool, exprValue any) (bool, error) {
...
+				)
+			}
+		}
+		return false, nil
+
+	default:
+		return false, NewErrorMsg(
+			ErrBadValue,
+			fmt.Sprintf(
+				`Invalid numerical type code: %v`, exprValue),
```
Comment (AlekSi) [inline internal/handlers/common/filter.go:620]: The same formatting issue I mention every time

File: internal/types/array.go [hunk 3]
```diff
@@ -128,3 +132,53 @@ func (a *Array) Append(values ...any) error {
 	a.s = append(a.s, values...)
 	return nil
 }
+
+// HasSameTypeElements returns true is Array elements has the same type.
```
Comment (AlekSi) [inline internal/types/array.go:136]:
```suggestion
// HasSameTypeElements returns true if Array elements has the same type.
```

File: internal/types/array.go [hunk 4]
```diff
@@ -128,3 +132,53 @@ func (a *Array) Append(values ...any) error {
...
+func (a *Array) HasSameTypeElements() bool {
+	var prev string
+	for _, i := range a.s {
+		var cur string
+		switch i := i.(type) {
+		case int32, int64:
```
Comment (AlekSi) [inline internal/types/array.go:142]: Let's use the same consistent order as everywhere: https://github.com/FerretDB/FerretDB/blob/6e8a7773a9867f76e982ed9229594d6c899e8c46/internal/types/types.go#L26-L44

File: internal/types/array.go (hunk 3 above)
Comment (AlekSi) [reply internal/types/array.go:136]: Also, it does not mentions the weird cases for number types

File: internal/handlers/common/params.go [hunk 5]
```diff
@@ -192,3 +192,47 @@ func getBinaryParams(fieldValue any, maskValue any) (types.Binary, types.Binary,
...
+	case 17:
+		return "timestamp", nil
+	case 18:
+		return "long", nil
+	case 19:
+		return "decimal", nil
+	case -1:
+		return "minKey", nil
+	case 127:
+		return "maxKey", nil
```
Comment (AlekSi) [inline internal/handlers/common/params.go:230]: I think we should not add codes for types we don't actually support

File: internal/handlers/common/params.go [hunk 6]
```diff
@@ -192,3 +192,47 @@ func getBinaryParams(fieldValue any, maskValue any) (types.Binary, types.Binary,
 
 	return fieldBinary, maskBinary, nil
 }
+
+// getTypeAliasByCode returns type alias and error by given type code.
```
Comment (AlekSi) [inline internal/handlers/common/params.go:196]: Somewhere should be a description what "type alias" and "type code" are.

One option could be to define custom types for them:
```go
// typeCode ...
type typeCode int32
```

It then allows us to define constants and their string representation by using stringer code generator

File: integration/query_element_test.go [hunk 7]
```diff
@@ -85,6 +89,180 @@ func TestExistsOperator(t *testing.T) {
...
+			err = cursor.All(ctx, &actual)
+			require.NoError(t, err)
+			assert.Equal(t, tc.expectedIDs, collectIDs(t, actual))
+		})
+	}
+}
+
+func TestTypeOperator(t *testing.T) {
+	t.Parallel()
+	// TODO: add this data types to collection "objectId", "decimal", "minKey", "maxKey"
```
Comment (AlekSi) [inline integration/query_element_test.go:102]: objectID was added, and I don't think we ever add minKey and maxKey

File: integration/query_element_test.go [hunk 8]
```diff
@@ -85,6 +89,180 @@ func TestExistsOperator(t *testing.T) {
...
+		},
+		"Timestamp": {
+			v:           "timestamp",
+			expectedIDs: []any{"timestamp", "timestamp-i"},
+		},
+		"Document": {
+			v:           "object",
+			expectedIDs: []any{"document", "document-empty"},
+		},
+		"Double": {
```
Comment (AlekSi) [inline integration/query_element_test.go:172]: The order of tests seems random

File: integration/query_element_test.go [hunk 9]
```diff
@@ -85,6 +89,180 @@ func TestExistsOperator(t *testing.T) {
 			require.NoError(t, err)
 
 			var actual []bson.D
+
```
Comment (AlekSi) [inline integration/query_element_test.go:92]: No blank line there

File: internal/handlers/common/params.go (hunk 5 above)
Comment (w84thesun) [reply internal/handlers/common/params.go:230]: Should I just remove unsupported type codes or add them to a reparate case that would return ErrNotImplemented or something like that?

Comment (AlekSi) [reply internal/handlers/common/params.go:230]: Being explicit with ErrNotImplemented is the best way, I think

File: integration/query_element_test.go [hunk 10]
```diff
@@ -85,6 +89,180 @@ func TestExistsOperator(t *testing.T) {
...
+			err = cursor.All(ctx, &actual)
+			require.NoError(t, err)
+			assert.Equal(t, tc.expectedIDs, collectIDs(t, actual))
+		})
+	}
+}
+
+func TestTypeOperator(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (w84thesun) [reply integration/query_element_test.go]: Fixed

File: internal/handlers/common/filter.go (hunk 2 above)
Comment (w84thesun) [reply internal/handlers/common/filter.go]: Fixed

File: internal/types/array.go (hunk 4 above)
Comment (w84thesun) [reply internal/types/array.go]: Fixed

File: integration/query_element_test.go (hunk 7 above)
Comment (w84thesun) [reply integration/query_element_test.go]: I've added test for objectID, removed minKey and maxKey from TODO.

File: integration/query_element_test.go (hunk 9 above)
Comment (w84thesun) [reply integration/query_element_test.go]: Fixed

File: integration/query_element_test.go (hunk 8 above)
Comment (w84thesun) [reply integration/query_element_test.go]: Rearranged tests as mentioned above.

**Review by seeforschauer — CHANGES_REQUESTED** [review]
what's left: `type aliases` and the function of getting the type resolved in the type itself when user input is string.

**Review by seeforschauer — DISMISSED** [review]

File: internal/types/array.go [hunk 11]
```diff
@@ -128,3 +132,58 @@ func (a *Array) Append(values ...any) error {
 	a.s = append(a.s, values...)
 	return nil
 }
+
+// HasSameTypeElements returns true if Array elements has the same type.
+// MongoDB consider int32, int64 and float64 that could be converted to int as the same type.
+func (a *Array) HasSameTypeElements() bool {
```
Comment (AlekSi) [inline internal/types/array.go:138]: That function has much more in common with unexported typeCode type and other type code / alias logic than with anything else in types package. Let's move it to `common` package

File: internal/handlers/common/filter.go [hunk 12]
```diff
@@ -511,3 +520,214 @@ func filterFieldExprExists(fieldExist bool, exprValue any) (bool, error) {
...
+
+		return res, nil
+
+	case *types.Array:
+		hasSameType := exprValue.HasSameTypeElements()
+
+		for i := 0; i < exprValue.Len(); i++ {
+			exprVal, err := exprValue.Get(i)
+			if err != nil {
+				panic(err)
```
[the remark this answers is not in the document]
Comment (w84thesun) [reply internal/handlers/common/filter.go]: Fixed

File: internal/handlers/common/filter.go [hunk 13]
```diff
@@ -511,3 +520,214 @@ func filterFieldExprExists(fieldExist bool, exprValue any) (bool, error) {
...
+			}
+
+			switch exprValue := exprVal.(type) {
+			case string:
```
[the remark this answers is not in the document]
Comment (w84thesun) [reply internal/handlers/common/filter.go]: Fixed

File: internal/handlers/common/filter.go [hunk 14]
```diff
@@ -511,3 +520,214 @@ func filterFieldExprExists(fieldExist bool, exprValue any) (bool, error) {
...
+	case *types.Array:
+		hasSameType := exprValue.HasSameTypeElements()
+
+		for i := 0; i < exprValue.Len(); i++ {
+			exprVal, err := exprValue.Get(i)
+			if err != nil {
+				panic(err)
+			}
+
+			switch exprValue := exprVal.(type) {
```
[the remark this answers is not in the document]
Comment (w84thesun) [reply internal/handlers/common/filter.go]: Fixed

File: internal/handlers/common/typecode.go [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package common
+
+import "fmt"
+
+//go:generate ../../../bin/stringer -linecomment -type typeCode
+
+// typeCode represents BSON type codes.
```
[the remark this answers is not in the document]
Comment (w84thesun) [reply internal/handlers/common/typecode.go:28]: PTAL, I tried to add some context.

File: internal/handlers/common/typecode.go [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+	case 19, -1, 127:
+		return 0, NewErrorMsg(ErrNotImplemented, fmt.Sprintf(`Type code %v not implemented`, code))
+	default:
+		return 0, NewErrorMsg(ErrBadValue, fmt.Sprintf(`Invalid numerical type code: %d`, code))
+	}
+}
+
+// aliasToTypeCode matches string type aliases to the corresponding typeCode value.
+var aliasToTypeCode = map[string]typeCode{
+	"double":    typeCodeDouble,
```
[the remark this answers is not in the document]
Comment (w84thesun) [reply internal/handlers/common/typecode.go]: Fixed

File: internal/types/array.go (hunk 11 above)
Comment (w84thesun) [reply internal/types/array.go]: Moved it to typecode.go and unexported it.

File: integration/query_element_test.go [hunk 17]
```diff
@@ -91,3 +95,185 @@ func TestExistsOperator(t *testing.T) {
...
+		},
+		"Timestamp": {
+			v:           "timestamp",
+			expectedIDs: []any{"timestamp", "timestamp-i"},
+		},
+		"Long": {
+			v:           "long",
+			expectedIDs: []any{"int64", "int64-max", "int64-min", "int64-zero"},
+		},
+
```
Comment (seeforschauer) [inline integration/query_element_test.go:165]: why have a blank line here?

File: internal/handlers/common/filter.go [hunk 18]
```diff
@@ -511,3 +520,214 @@ func filterFieldExprExists(fieldExist bool, exprValue any) (bool, error) {
 		return false, nil
 	}
 }
+
+// filterFieldExprType handles {field: {$type: value}} filter.
+func filterFieldExprType(fieldValue, exprValue any) (bool, error) {
+	switch exprValue := exprValue.(type) {
+	case string:
```
[the remark this answers is not in the document]
Comment (seeforschauer) [reply internal/handlers/common/filter.go]: > Please use our standard order of types

@AlekSi would you be so kind to let us know where can we read about our standard order of types? Just searched in the repository and didn't found any.

**Review by seeforschauer — APPROVED** [review]

File: internal/handlers/common/filter.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (AlekSi) [reply internal/handlers/common/filter.go]: https://github.com/FerretDB/FerretDB/blob/6ada63ed03c90919c11a0314962def6e441977cf/internal/types/types.go#L26-L44

It is the same as at https://bsonspec.org/spec.html

**Comment by AlekSi** [comment]
(re-enabled with the right commit body)

File: integration/query_element_test.go (hunk 17 above)
Comment (w84thesun) [reply integration/query_element_test.go:165]: To separate `number` case

**Review by AlekSi — APPROVED** [review]

**Pull request merged**